### ISSUE TITLE
Approve/Decline images using GPIO/keyboard

### ISSUE THREAD

**Issue by mgineer85** [photobooth-app/photobooth-app#250]
Allow more flexible configuration of gpio and keyboard input to approve/decline images.

**Comment by mgineer85** [comment]
reference to another post: https://github.com/photobooth-app/photobooth-app/issues/203#issuecomment-2199200479

**Comment by McKaiver** [comment]
Yes, as i said in the other topic, It would be nice, that you could controll the box with a few buttons.

Touch screens have a big dissadvantage in "Public" areas, because everybody touches it, even with dirty fingers and then there is sauce on the screen and so on, especially, because this device will be used at a party, with drunk people.

**Comment by mgineer85** [comment]
So you have 3 buttons.    What are these buttons labeled like to the user?

A: Caputure Image / Approve(?)
B: Reject(?)
C: Print(?)

The api endpoints used by the touchscreen have 3 additional routes: confirm+next, reject, abort. 
The GPIO service (and for equality the keyboard input then also in the end) needs to have similar, like:
```

    def _handle_processing_next_confirm_button(self):
        try:
            self._processing_service.continue_process()
            return "OK"
        except Exception as exc:
            # other errors
            logger.critical(exc)

    def _handle_processing_reject_button(self):
        try:
            self._processing_service.reject_capture()
            return "OK"
        except Exception as exc:
            # other errors
            logger.critical(exc)

    def _handle_processing_abort_button(self):
        try:
            self._processing_service.abort_process()

        except Exception as exc:
            # other errors
            logger.critical(exc)
```

Hardest part will be the matching between the buttons and the function if the GPIO pin is already occupied by another function (like confirm and capture on one button).

**Comment by mgineer85** [comment]
There are changes to the GPIO handling now made in the dev branch.

There are 3 additional gpio inputs now that can be defined:
- next
  - is to confirm the capture
  - or for a video to stop capture
- reject
  - is to reject the capture and repeat it
  - or for a video it stops the capture (but it's not rejected, just the capture is stopped and the result presented)
- abort: aborts the job completely.

With the changes added, it is possible to reuse a GPIO pin. That allows to use 2 buttons that share logic.
If no user input is required (means, no approval state), the input next/reject/abort is ignored. If the GPIO has an action assigned it will be triggered - for example create a collage.

If a user input would be required in the moment the gpio is active, it will ignore the action but continue as requested by the user (approve/reject/abort).

Hope that makes sense?

Would be great if somebody else would test the dev branch to confirm it's working as intended...

![Image](https://github.com/user-attachments/assets/216eacc8-303f-4c1a-a65a-997d627b8bce)

**Comment by BermudaStudio** [comment]
Hi,
Is it possible to assign keyboard inputs for these next/reject/abort actions ?
(I would like to use it with a PC without GPIO)

**Comment by mgineer85** [comment]
Hi, keyboard inputs cannot trigger next/reject/abort currently.    How would the keyboard (or other input device) look like for this kind of application? I guess it cannot be a standard HMI because it would look a bit messy?   GPIO are not acceptable for you?

**Comment by BermudaStudio** [comment]
To build a vintage photobooth cabin without touchscreen, I use a 2014 Mac Mini with a Joypad USB card that emulates key presses from arcade buttons (thanks to https://github.com/sezanzeb/input-remapper ).
The title of this issue gave me hope that this was possible. 
As TAB key highlights alternately each active link on Chromium and ENTER acts like a click on it, I will try this solution : One button emulates TAB key and another emulates ENTER key, with instructions like "button 1 to select, button 2 to validate your choice).
All I have to do is figure out how to amplify link highlighting with CSS…
